File: insertion/insertion_parallel_new.cpp
#include <algorithm>
#include <vector>

void parallelInsertionSort(std::vector<int>& arr, int left, int right) {
    if (left >= right) {
        return;
    }

    const int n = right - left + 1;
    std::vector<int> output(static_cast<std::size_t>(n));

#ifdef _OPENMP
#pragma omp parallel for schedule(static)
#endif
    for (int offset = 0; offset < n; ++offset) {
        const int i = left + offset;
        const int value = arr[static_cast<std::size_t>(i)];
        int rank = 0;

        for (int j = left; j <= right; ++j) {
            const int other = arr[static_cast<std::size_t>(j)];
            if (other < value || (other == value && j < i)) {
                ++rank;
            }
        }

        output[static_cast<std::size_t>(rank)] = value;
    }

    std::copy(output.begin(), output.end(), arr.begin() + left);
}

File: insertion/insertion_parallel_new.jl
using Polyester

function parallel_insertion_sort!(arr::Vector)
    n = length(arr)
    n < 2 && return arr

    output = similar(arr)
    @batch for i in 1:n
        value = arr[i]
        rank = 1
        for j in 1:n
            other = arr[j]
            if other < value || (other == value && j < i)
                rank += 1
            end
        end
        output[rank] = value
    end

    copyto!(arr, output)
    return arr
end
